src/dependency_manager/manager.py

```diff
@@ -6,14 +6,25 @@
 
 
 class DependencyManagerAbstract(Singleton, ABC):
-    def init(self, dry_run=False):
+    def init(self):
         """One-time class initialization."""
         self.parent_directory = self.get_parent_dir()
-        self.dependency_file = self._infer_dependency_files()
-        self.dependencies = self._infer_dependencies()
+        self._dependency_file = None
+        self._dependencies = None
         self.dependencies_changed = False
         self.dependency_file_changed = False
-        self.dry_run = dry_run
+
+    @property
+    def dependency_file(self):
+        if not self._dependency_file:
+            self._dependency_file = self._infer_dependency_files()
+        return self._dependency_file
+
+    @property
+    def dependencies(self):
+        if not self._dependencies:
+            self._dependencies = self._infer_dependencies()
+        return self._dependencies
```

Comment:
you can actually remove these new impl and just make the original methods into cached_properties https://docs.python.org/3/library/functools.html#functools.cached_property

HOWEVER... .I'm actually not sure we need this, given that the point of `init` in a singleton is to ever run once. So if we only call these properties / methods in init, they're already cached. Am I mistaken?

Reply:
The point is being lazy, `_infer_dependencies` will only be called if `depedencies` is accessed (essentially when `add` or `remove` is called).

For example, if we have
```python
DependencyManager().write()
```
and nothing else, then `dependencies_changed == False` as per the `init()`. This means that  `_infer_dependencies()` will never be called since `dependencies` property is never accessed within `write` due to the check at line 65. If we had it at `init()`, then `_infer_dependencies()` would be called regardless.

Good point about cached property. Will test if lazyness is preserved and make changes (will save us the `if`).